File: apps/frontend/src/lib/contracts/payment.ts
import { z } from "zod"

export const ConsentRequestSchema = z.object({
  schema_version: z.literal("1.0").default("1.0"),
  request_id: z.string().min(1),
  consent_to_marketing: z.boolean(),
  terms_accepted: z.literal(true),
  data_processing_consent: z.literal(true),
})

export const ConsentResponseSchema = z.object({
  schema_version: z.literal("1.0"),
  request_id: z.string(),
  consent_recorded: z.boolean(),
  message: z.string(),
})

export const PaymentSubmitRequestSchema = z.object({
  schema_version: z.literal("1.0").default("1.0"),
  request_id: z.string().min(1),
  amount: z.number().int().positive(),
  method: z.enum(["LOCAL_BANK_TRANSFER", "CASH", "STRIPE", "PAYPAL"]),
})

export const PaymentSubmitResponseSchema = z.object({
  schema_version: z.literal("1.0"),
  payment_id: z.string(),
  request_id: z.string(),
  amount: z.number(),
  method: z.string(),
  status: z.string(),
  payment_instructions: z.string().nullable(),
  expected_wait_time: z.string().nullable(),
})

export const PaymentStatusResponseSchema = z.object({
  schema_version: z.literal("1.0"),
  payment_id: z.string(),
  request_id: z.string(),
  amount: z.number(),
  method: z.string(),
  status: z.string(),
  created_at: z.string(),
  verified_at: z.string().nullable(),
  ocr_extracted: z.record(z.unknown()).nullable(),
  admin_notes: z.string().nullable(),
})

export const CampaignExecuteRequestSchema = z.object({
  schema_version: z.literal("1.0").default("1.0"),
  request_id: z.string().min(1),
  campaign_data: z.record(z.unknown()),
})

export const CampaignExecuteResponseSchema = z.object({
  schema_version: z.literal("1.0"),
  execution_id: z.string(),
  request_id: z.string(),
  status: z.string(),
  message: z.string(),
})

export type ConsentRequest = z.infer<typeof ConsentRequestSchema>
export type ConsentResponse = z.infer<typeof ConsentResponseSchema>
export type PaymentSubmitRequest = z.infer<typeof PaymentSubmitRequestSchema>
export type PaymentSubmitResponse = z.infer<typeof PaymentSubmitResponseSchema>
export type PaymentStatusResponse = z.infer<typeof PaymentStatusResponseSchema>
export type CampaignExecuteRequest = z.infer<typeof CampaignExecuteRequestSchema>
export type CampaignExecuteResponse = z.infer<typeof CampaignExecuteResponseSchema>

File: apps/backend/app/models/payment.py
import re
from datetime import datetime
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator


class PaymentMethod(str, Enum):
    LOCAL_BANK_TRANSFER = "LOCAL_BANK_TRANSFER"
    CASH = "CASH"
    STRIPE = "STRIPE"
    PAYPAL = "PAYPAL"


class PaymentStatus(str, Enum):
    PENDING = "PENDING"
    APPROVED = "APPROVED"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    REFUNDED = "REFUNDED"


class ExecutionStatus(str, Enum):
    PENDING = "PENDING"
    RUNNING = "RUNNING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"


EMAIL_PATTERN = re.compile(r"^[^@\s]+@[^@\s]+\.[^@\s]+$")


class PaymentSubmitRequest(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    request_id: str = Field(min_length=1)
    amount: int = Field(ge=1)
    method: PaymentMethod
    auto_approve: bool = False

    @field_validator("request_id")
    @classmethod
    def validate_request_id(cls, value: str) -> str:
        if not value.startswith("REQ-"):
            raise ValueError("request_id must start with 'REQ-'")
        return value


class PaymentSubmitResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    payment_id: str
    request_id: str
    amount: int
    method: PaymentMethod
    status: PaymentStatus
    payment_instructions: Optional[str] = None
    expected_wait_time: Optional[str] = None
    stripe_client_secret: Optional[str] = None
    message: Optional[str] = None


class PaymentVerifyRequest(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    payment_id: str
    action: Literal["approve", "reject"]
    admin_notes: Optional[str] = None


class PaymentVerifyResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    payment_id: str
    request_id: str
    status: PaymentStatus
    message: str
    notification_sent: bool = False


class PaymentStatusResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    payment_id: str
    request_id: str
    amount: int
    method: PaymentMethod
    status: PaymentStatus
    created_at: datetime
    verified_at: Optional[datetime] = None
    ocr_extracted: Optional[dict] = None
    admin_notes: Optional[str] = None
    audit_trail: Optional[list[dict]] = None


class OCRExtractedData(BaseModel):
    amount: Optional[int] = None
    date: Optional[str] = None
    reference_number: Optional[str] = None
    sender_name: Optional[str] = None
    sender_account: Optional[str] = None
    confidence: Optional[float] = None


class ConsentRecord(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    request_id: str
    consent_to_marketing: bool
    terms_accepted: bool
    data_processing_consent: bool
    consented_at: datetime = Field(default_factory=datetime.utcnow)


class ConsentRequest(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    request_id: str
    consent_to_marketing: bool
    terms_accepted: bool
    data_processing_consent: bool

    @field_validator("terms_accepted")
    @classmethod
    def validate_terms(cls, value: bool) -> bool:
        if not value:
            raise ValueError("terms_accepted must be true")
        return value


class ConsentResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    request_id: str
    consent_recorded: bool
    message: str


class CampaignExecuteRequest(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    request_id: str
    campaign_data: dict


class CampaignExecuteResponse(BaseModel):
    schema_version: Literal["1.0"] = "1.0"
    execution_id: str
    request_id: str
    status: ExecutionStatus
    message: str


class ExecutionJob(BaseModel):
    id: str
    request_id: str
    status: ExecutionStatus
    adapter_type: str
    executed_at: Optional[datetime] = None
    result_data: Optional[dict] = None